compiler/resolution/lowerIterators.cpp

```diff
@@ -1581,40 +1579,66 @@ static void processShadowVariables(ForLoop* forLoop, SymbolMap *map) {
+          // outerVar
+          if (DefExpr* prevDef = toDefExpr(svar->defPoint->prev)) {
+            if (ShadowVarSymbol* castTemp = toShadowVarSymbol(prevDef->sym)) {
+              if (castTemp->isCompilerAdded()) {
+                Symbol* castOuter = castTemp->outerVarSE->symbol();
+                if (castOuter->hasFlag(FLAG_TFI_BORROW_TEMP)) {
+                  outerVarSym = castOuter;
+                }
               }
             }
           }
-        }
-
-        map1.put(svar->ParentvarForIN(), outerVarSym);
-        Expr *copiedInitialization = initMove->get(2)->copy(&map1);
-        forLoop->insertBefore(new CallExpr(PRIM_MOVE, capturedSvar, copiedInitialization));
 
-        VarSymbol* taskIndVar = new VarSymbol(astr("taskInd_", svar->name), svar->type);
-        taskIndVar->addFlag(FLAG_TASK_PRIVATE_VARIABLE);
-        forLoop->insertBefore(new DefExpr(taskIndVar));
-        forLoop->insertBefore(new CallExpr(
-            PRIM_MOVE, taskIndVar,
-            new CallExpr(PRIM_TASK_PRIVATE_SVAR_CAPTURE, copiedInitialization->copy())));
-
-        map->put(svar, taskIndVar);
+          SymbolMap mapForInitCopy;
+          mapForInitCopy.put(svar->ParentvarForIN(), outerVarSym);
+          Expr *copiedInitialization = initMove->get(2)->copy(&mapForInitCopy);
+          VarSymbol* taskIndVar = new VarSymbol(astr("taskInd_", svar->name), svar->type);
+          taskIndVar->addFlag(FLAG_TASK_PRIVATE_VARIABLE);
+          forLoop->insertBefore(new DefExpr(taskIndVar));
+          forLoop->insertBefore(new CallExpr(
+              PRIM_MOVE, taskIndVar,
+              new CallExpr(PRIM_TASK_PRIVATE_SVAR_CAPTURE, copiedInitialization)));
+
+          map->put(svar, taskIndVar);
+        } else {
+          // If the initialization block doesn't use a MOVE expression but rather
+          // calls an init function directly (passing the to-be initialized object
+          // in by ref) then we process that differently.
+          //
+          // IOW we are given:
+          //   (CallExpr
+          //      (fn init =)
+          //      (val x)
+          //      (INP_x))
+          //
+          // And we want:
+          //
+          //   PRIM_TASK_PRIVATE_SVAR_CAPTURE(init=(taskInd_x, capX));
+          VarSymbol* taskIndVar = new VarSymbol(astr("taskInd_", svar->name), svar->type);
+          taskIndVar->addFlag(FLAG_TASK_PRIVATE_VARIABLE);
+          forLoop->insertBefore(new DefExpr(taskIndVar));
+
+          SymbolMap mapForInitCopy;
+          Symbol* outerVarSym = svar->outerVarSE->symbol();
+          mapForInitCopy.put(svar->ParentvarForIN(), outerVarSym);
+          mapForInitCopy.put(svar, taskIndVar);
+          Expr *copiedInitialization = initMove->copy(&mapForInitCopy);
+          forLoop->insertBefore(
+              new CallExpr(PRIM_TASK_PRIVATE_SVAR_CAPTURE, copiedInitialization));
+
+          map->put(svar, taskIndVar);
+        }
```

Comment:
This is hard to understand looking at the diff, but this brace and the one below are at the same indentation level.